_posts/2018-07-20-totw-153.md

```diff
@@ -0,0 +1,211 @@
+namespace example {
+namespace {
+
+TEST(MyTest, UsesUsingDirectives) {
+  Sequence seq; // ::testing::Sequence
+  WallTimer timer; // ::WallTimer
+  ...
+}
+
+} // namespace
+} // namespace example
+} // namespace totw
+```
+
+Let's see what kind of changes can break this code:
+
+*   If anyone defines `::totw::Sequence` or `::totw::example::Sequence`, `seq`
+    will now refer to that entity instead of `::testing::Sequence`.
+*   If anyone defines `::Sequence`, the definition of `seq` will fail to
+    compile, as the reference to the name `Sequence` will be ambiguous.
+    `Sequence` could mean `::testing::Sequence` or `::Sequence`, and the
+    compiler doesn't know which one you wanted.
+*   If anyone defines `::testing::WallTimer`, the definition of `timer` will
+    fail to compile.
+
+Thus, a single using-directive in a function scope has placed naming
+restrictions on symbols in `::testing`, `::totw`, `::totw::example`, and the
+global namespace. **Allowing this using-directive, even if only in function
+scope, creates ample opportunities for name clashes in the global and other
+namespaces.**
+
+If that example doesn't look fragile enough, consider this:
+
+```c++
+namespace totw {
+namespace example {
+namespace {
+
+TEST(MyTest, UsesUsingDirectives) {
+  using namespace ::testing;
+  EXPECT_THAT(..., proto::Partially(...)); // ::testing::proto::Partially
+  ...
+}
+
+} // namespace
+} // namespace example
+} // namespace totw
+```
+
+This using-directive has basically **introduced a namespace alias** `proto` in
+the global namespace, roughly equivalent to the following:
+
+<pre class="prettyprint lang-cpp bad-code">
+namespace proto = ::testing::proto;
+
+namespace totw {
+namespace example {
+namespace {
+
+TEST(MyTest, UsesUsingDirectives) {
+  EXPECT_THAT(..., proto::Partially(...)); // ::testing::proto::Partially
+  ...
+}
+
+} // namespace
+} // namespace example
+} // namespace totw
+</pre>
+
+The test will keep compiling until a header defining namespace `::proto`,
+`::totw::proto`, or `::totw::example::proto` gets included transitively. At that
+point in time, `proto::Partially` becomes ambiguous, and the test stops
+compiling. This ties into the style guide's rules on namespace naming: avoid
+nested namespaces, and don't use common names for nested namespaces. (See [TotW
+130](http://abseil.io/tips/130) and the style rules on [Namespace
+names](https://google.github.io/styleguide/cppguide.html#Namespace_Names) for
+more on this topic.)
+
+One might think that it's safe to employ a using-directive for a closed
+namespace that has few symbols and guarantees that no more symbols will be ever
+added to it. (`std::placeholders`, which contains symbols `_1` ... `_9`, is an
+example of such a namespace.) However, even that isn't safe: it precludes any
+other namespace from introducing symbols with the same names. In this sense,
+using-directives defeat the modularity provided by namespaces.
+
+## Unqualified using-directives
```

Comment:
This section is longer and more complex than any other and it doesn't seem to add much. Maybe delete it?